[mongo] (optional) `additional_metrics` parameter

yannmh opened the pull request:
Extends #1903 and #1980

**More changes:**
By default, the check collects a sample of metrics from MongoDB.
The (optional) `additional_metrics` parameter instructs the check to collect additional
metrics on specific topics.
Available options are:
* `durability` - Journaling-related operations and performance
* `locks` - Locks
* `metrics.commands` - Use of database commands
* `tcmalloc` -  TCMalloc memory allocator
* `wiredTiger` - WiredTiger storage engine

Thanks again @scottbessler

benmccann commented:
Do you want to close my PR https://github.com/DataDog/dd-agent/pull/1980 and include those changes here? There seems to be a lot of overlap (one is rebased on the other?), so it's little tough to review both. Might as well just do one PR for the additional metrics?

benmccann commented on checks.d/mongo.py:
```diff
@@ -101,21 +119,100 @@ class MongoDb(AgentCheck):
+
+        https://docs.mongodb.org/manual/reference/command/serverStatus/#server-status-locks
+        """
+        "locks.Collection.acquireCount.R": RATE,
```
The metrics you're reporting throughout don't quite add up with what I see. It could be a difference between MMAP and WiredTiger (I'm running WiredTiger and am guessing you did this with MMAP?)

Ones I see that you're missing:

    "locks.Collection.acquireCount.r",
    "locks.Collection.acquireCount.w",
    "locks.Database.acquireCount.R",
    "locks.Database.acquireCount.W",
    "locks.Database.acquireWaitCount.R",
    "locks.Database.acquireWaitCount.W",
    "locks.Database.timeAcquiringMicros.R",
    "locks.Database.timeAcquiringMicros.W",
    "locks.Global.acquireCount.R",
    "locks.Global.acquireCount.W",
    "locks.Global.acquireWaitCount.R",
    "locks.Global.acquireWaitCount.W",
    "locks.Global.timeAcquiringMicros.R",
    "locks.Global.timeAcquiringMicros.W",

    "network.bytesIn,
    "network.bytesOut",
    "network.numRequests",

    "metrics.commands._isSelf.failed",
    "metrics.commands._isSelf.total",
    "metrics.commands.aggregate.failed",
    "metrics.commands.aggregate.total",
    "metrics.commands.buildInfo.failed",
    "metrics.commands.buildInfo.total",
    "metrics.commands.collStats.failed",
    "metrics.commands.collStats.total",
    "metrics.commands.dbStats.failed",
    "metrics.commands.dbStats.total",
    "metrics.commands.distinct.failed",
    "metrics.commands.distinct.total",
    "metrics.commands.getLastError.failed",
    "metrics.commands.getLastError.total",
    "metrics.commands.getLog.failed",
    "metrics.commands.getLog.total",
    "metrics.commands.getnonce.failed",
    "metrics.commands.getnonce.total",
    "metrics.commands.isMaster.failed",
    "metrics.commands.isMaster.total",
    "metrics.commands.listCollections.failed",
    "metrics.commands.listCollections.total",
    "metrics.commands.listIndexes.failed",
    "metrics.commands.listIndexes.total",
    "metrics.commands.ping.failed",
    "metrics.commands.ping.total",
    "metrics.commands.renameCollection.failed",
    "metrics.commands.renameCollection.total",
    "metrics.commands.replSetGetStatus.failed",
    "metrics.commands.replSetGetStatus.total",
    "metrics.commands.replSetHeartbeat.failed",
    "metrics.commands.replSetHeartbeat.total",
    "metrics.commands.replSetUpdatePosition.failed",
    "metrics.commands.replSetUpdatePosition.total",
    "metrics.commands.saslContinue.failed",
    "metrics.commands.saslContinue.total",
    "metrics.commands.whatsmyuri.failed",
    "metrics.commands.whatsmyuri.total",
    "metrics.operation.writeConflicts",
    "metrics.repl.buffer.count",
    "metrics.repl.buffer.maxSizeBytes",
    "metrics.repl.buffer.sizeBytes",
    "metrics.repl.preload.docs.num",
    "metrics.repl.preload.docs.totalMillis",
    "metrics.repl.storage.freelist.search.bucketExhausted",
    "metrics.repl.storage.freelist.search.requests",
    "metrics.repl.storage.freelist.search.scanned",

yannmh commented:
Done. 
I also added more changes to the PR.

**Support for custom metric name aliases**
MongoDB metric names are now defined as follow:
* If available, the normalized metric name alias
* (Or) the normalized original metric name

**Case-insensitive metric names**
Metric name normalization mechanism replaces case-sensitive metric name
suffixes.

**WiredTiger metric name aliases**
Renamed/Shrink a few WiredTiger metric names. 

@benmccann do you reckon some WiredTiger metrics should be reported as `rate` rather than `gauge` ?

Thanks!

benmccann commented:
@yannmh thanks for the help on this. I'm going to be super excited when we get this in

Please see my comment on one of the precursor PRs which I believe is still relevant here: https://github.com/DataDog/dd-agent/pull/1903#discussion_r43459192

benmccann commented:
Regarding whether some WiredTiger metrics should be reported as `rate` instead of `gauge`... 

I think the following metrics currently in the PR are monotonically increasing counters (i.e. start at 0 when mongodb is restarted and only increase from there):
wiredTiger.cache.failed eviction of pages that exceeded the in-memory maximum
wiredTiger.cache.failed_eviction_of_pages_exceeding_the_in-memory_maximum
wiredTiger.cache.pages evicted because they exceeded the in-memory maximum
wiredTiger.cache.pages evicted by application threads

If DataDog is capable of taking a monotonically increasing counter and transforming it to a `rate` then it'd make sense to change those.

benmccann commented:
@yannmh can you rebase this PR?

yannmh commented:
Thanks for the extra information @benmccann.
I am rebasing this PR, while bringing the necessary change to report the metrics you mentioned with the appropriate type  (`rate` instead of `gauge`).

About [ #1903 (comment)](https://github.com/DataDog/dd-agent/pull/1903#discussion_r43459192), I see that you are listing multiple metrics that are not currently collected. Do you think all of them worth being added the collection scope ? It would be nice to keep the list of metrics succinct to target "core"/"significant" metrics only.

benmccann commented:
@yannmh I think that mongo exports a very different set of metrics depending on whether you're using the default mmap storage engine or the newer wiredtiger storage engine. I'm not sure the number of metrics is that different, but the actual set of metrics varies quite a bit. So while I listed quite a few metrics that dd-agent isn't collecting, the current code also has quite a large number of metrics that I'm not seeing on my mongo instance. I think if you add all the metrics I listed it only makes people running wiredtiger collect as many metrics you're already collecting for the majority of users.

benmccann commented:
@yannmh just a quick reminder to rebase this PR. thanks!

yannmh commented:
Rebased on top of master. Also added a few additional changes based on [#1903(comment)](https://github.com/DataDog/dd-agent/pull/1903#discussion_r43459192), i.e. 

**Additional metrics within the following topics:**
    * Locks
    * Network
    * Operation
    * Replication
See 882ef98926ae6a2da7f5b1029d78223121166e2c

@benmccann, the check now covers the metrics you listed in [#1903(comment)](https://github.com/DataDog/dd-agent/pull/1903#discussion_r43459192). 
I omitted the `metrics.commands` ones for now : would you mind opening a PR, so we can more conveniently discuss about it ?  Specifically, I'd like to know how you use these metrics to monitor MongoDB instances :mag: .
Thanks again !

benmccann commented:
Ok, sounds good. Let's ship it!

yannmh commented:
Closing in favor of  #2161.
